lib/rules/max-lines-per-function.js

```diff
@@ -0,0 +1,219 @@
+    const map = new Map();
+
+    if (!comments) {
+        return map;
+    }
+    comments.forEach(comment => {
+        for (let i = comment.loc.start.line; i <= comment.loc.end.line; i++) {
+            map.set(i, comment);
+        }
+    });
+    return map;
+}
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "enforce a maximum number of line of code in a function",
+            category: "Stylistic Issues",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/max-lines-per-function"
+        },
+
+        schema: [
+            OPTIONS_OR_INTEGER_SCHEMA
+        ]
+    },
+
+    create(context) {
+        const sourceCode = context.getSourceCode();
+        const lines = sourceCode.lines;
+
+        const option = context.options[0];
+        let maxLines = 50;
+        let skipComments = false;
+        let skipBlankLines = false;
+        let IIFEs = false;
+
+        if (typeof option === "object") {
+            if (typeof option.max === "number") {
+                maxLines = option.max;
+            } else if (typeof option.maximum === "number") {
+                maxLines = option.maximum;
+            }
+            if (typeof option.skipComments === "boolean") {
+                skipComments = option.skipComments;
+            }
+            if (typeof option.skipBlankLines === "boolean") {
+                skipBlankLines = option.skipBlankLines;
+            }
+            if (typeof option.IIFEs === "boolean") {
+                IIFEs = option.IIFEs;
+            }
+        } else if (typeof option === "number") {
+            maxLines = option;
+        }
+
+        const commentLineNumbers = getCommentLineNumbers(sourceCode.getAllComments());
+
+        //--------------------------------------------------------------------------
+        // Helpers
+        //--------------------------------------------------------------------------
+
+        /**
+         * Tells if a comment encompasses the entire line.
+         * @param {string} line The source line with a trailing comment
+         * @param {number} lineNumber The one-indexed line number this is on
+         * @param {ASTNode} comment The comment to remove
+         * @returns {boolean} If the comment covers the entire line
+         */
+        function isFullLineComment(line, lineNumber, comment) {
+            const start = comment.loc.start,
+                end = comment.loc.end,
+                isFirstTokenOnLine = !line.slice(0, comment.loc.start.column).trim();
+
+            return comment &&
+                (start.line < lineNumber || (start.line === lineNumber && isFirstTokenOnLine)) &&
+                (end.line > lineNumber || (end.line === lineNumber && end.column === line.length));
```

Comment:
Will this be tripped up by a block comment with trailing whitespace? (e.g., `/* some comment */\t\t` where `\t` is a tab stop, or any other trailing whitespace)

Reply:
You're absolutely right - this function was copied verbatim from the ```max-len``` rule, so that will have the same issue too.

Reply:
I've checked the ```max-len``` rule to see if it's got the same problem, but it's ok because they trim the end of the line previously. I've fixed this issue here though